ichnaea/locate/provider.py

```diff
@@ -633,24 +654,35 @@ def locate(self, data):
                     verify=False,
                 )
             self.stat_count('fallback.lookup_status.%s' % response.status_code)
-            if response.status_code == 404:
+            if response.status_code != 404:
                 # don't log exceptions for normal not found responses
-                response = None
-            else:
                 response.raise_for_status()
+
+            return response.json()
+
         except requests.exceptions.RequestException:
             self.raven_client.captureException()
-            response = None
 
-        if response:
-            try:
-                json_data = response.json()
-                location = self.location_type(
-                    lat=json_data['location']['lat'],
-                    lon=json_data['location']['lng'],
-                    accuracy=json_data['accuracy'],
-                )
-            except (KeyError, TypeError):
-                self.raven_client.captureException()
+    def locate(self, data):
+        location = self.location_type(query_data=False)
+
+        if not self.limit_reached():
+
+            location_data = (
+                self._fetch_cached_result(data) or
+                self._make_external_call(data)
+            )
+
+            if location_data:
+                try:
+                    location = self.location_type(
+                        lat=location_data['location']['lat'],
+                        lon=location_data['location']['lng'],
+                        accuracy=location_data['accuracy'],
+                    )
+                except (KeyError, TypeError):
+                    self.raven_client.captureException()
+
+                self._set_cached_result(data, location_data)
```

Comment:
Wouldn't this re-set the cached value we got earlier? So basically extending the expiry time on the cached value every time? It would also prevent us from actually checking the external service again for better data.

I'd only set the value here, if it didn't come from our cache.